include/drivers/can.h

```diff
@@ -167,34 +167,49 @@ struct can_filter {
  *
  */
 struct zcan_frame {
-	/** Indicates the identifier type (standard or extended)
-	 * use can_ide enum for assignment
-	 */
-	uint32_t id_type : 1;
+	/** Message identifier*/
+	uint32_t id      : 29;
+	/** Frame is in the CAN-FD frame format */
+	uint32_t fd      : 1;
 	/** Set the message to a transmission request instead of data frame
 	 * use can_rtr enum for assignment
 	 */
 	uint32_t rtr     : 1;
-	/** Message identifier*/
-	union {
-		uint32_t std_id  : 11;
-		uint32_t ext_id  : 29;
-	};
+	/** Indicates the identifier type (standard or extended)
+	 * use can_ide enum for assignment
+	 */
+	uint32_t id_type : 1;
 	/** The length of the message (max. 8) in byte */
 	uint8_t dlc;
-	/** The message data*/
-	union {
-		uint8_t data[8];
-		uint32_t data_32[2];
-	};
+	/** Baud Rate Switch. Frame transfer with different timing during
+	 * the data phase. Only valid for CAN-FD
+	 */
+	uint8_t brs : 1;
+	/** If set, buf points to the data, else data hold the data */
+	uint8_t ext_buf : 1;
+	/** Reserved for future flags */
+	uint8_t res : 6;
 #if defined(CONFIG_CAN_RX_TIMESTAMP)
-	/** Timer value of the CAN free running timer.
+	/** Timer value of the CAN free-running timer.
 	 * The timer is incremented every bit time and captured at the start
 	 * of frame bit (SOF).
 	 */
 	uint16_t timestamp;
+#else
+	/** @cond INTERNAL_HIDDEN */
+	uint8_t res0;  /* reserved / padding */
+	uint8_t res1;  /* reserved / padding */
+	/** @endcond */
 #endif
-} __packed;
+	/** The frame payload data. */
+	union {
+		/** frame data, if ext_buf is unset */
+		uint8_t data[CAN_MAX_DLEN];
+		uint32_t data_32[ceiling_fraction(CAN_MAX_DLEN, sizeof(uint32_t))];
+		/** pointer ot the frame data, if ext_buf is set */
+		const uint8_t *buf;
+	};
```

Comment:
Should `CAN_MAX_DLEN` become Kconfig-urable for CAN-FD?

As mentioned in the issue, I'm not a huge fan of unions (MISRA even forbids them, btw.). With up to 64 bytes in CAN-FD we would need `memcpy` instead of direct `uint32` copying anyway. So I would probably just use the most simple (and endian-safe) approach: `uint8_t data[CONFIG_CAN_FRAME_MAX_DLEN];`

I do see the benefit of using a pointer instead of the buffer to avoid copy operations in longer streams of data (e.g. for CAN-FD or ISO-TP). However, if each driver has to support both buffer and pointer frame types, this can create quite a lot of noise in the code. If we sacrifice 4 more bytes, we could do something like this (if we ensure that `data` points to the right location during initialization):

```C
#ifdef CONFIG_CAN_FRAME_INT_BUF
	uint8_t data_buf[CONFIG_CAN_FRAME_MAX_DLEN];
#endif
	const uint8_t *data;
```
The drivers could just always use the `data` pointer. Not totally sure if that makes sense...

Reply:
I would not add anything after the data, because we would lose the ability to cast shorter frames to z_can frames.
data_buf should always be present because receiving frames will always put the data in there.
We could still drop that commit.
Regarding the uint32_t data_32, I'd like to keep them, because otherwise, we would need to use casts in the drivers, which I generally don't like.

Reply:
Yes, CAN_MAX_DLEN is for CAN-FD

Reply:
I'm fine with the union to allow all different options to store the message. But don't we have to rewrite all drivers to support ext_buf? So this might be something for a separate PR.

Reply:
Removed that commit.

Reply:
Removed.